Add integration tests

roback opened the pull request:
Tested it with freebsd-mysql vagrant machine from ansible.

Requires the environment variables `DATABASE_URL`, `DATABASE_USER` and `DATABASE_PASS` because the JDBC adapter for sequel didn't like `username:password` inside the url for some reason :confused:.
- Add #stop method to Ecco::Client to be able to stop it when testing.
- Two helper classes:
  - DatabaseHelper - Inserts/updates/deletes table rows
  - TestHelper - Starts/stops ecco and returns the received row event
- Integration tests which checks that #on_row_event receives the correct type.
### Todo
- [x] MySQL with replication enabled on Travis
- [x] Vagrantfile
- [x] Documentation
- [x] Do not run integration tests by default
- [x] More tests
  - [x] Ensure that database/table information is correct
  - [x] Ensure that we can handle file rotations
  - [x] Set and get public properties
- [x] Remove duplicated code if possible
- [x] ~~Test for `client.stop`, `client.start` and `client.start_in_thread` (they are already tested implicitly in most of the tests)~~

close #2

roback committed:
Add integration test base

* Add #stop method to Ecco::Client to be able to stop it when testing.
* Two helper classes:
  * DatabaseHelper - Inserts/updates/deletes table rows
  * TestHelper - Starts/stops ecco and returns event
* Integration tests which checks that #on_row_event receives the
  correct type.

roback committed:
Enable mysql binary logging on travis

roback commented:
The tests passed on travis on the first try, that has never happened before :sweat_smile:

dentarg commented on spec/test_helper.rb:
```diff
@@ -0,0 +1,21 @@
+class TestHelper
+  def self.get_first_row_event_from_client(ecco_client)
```
What about `get_first_row_event_from_client` -> `start_client_get_first_row_event_then_stop_client`? :)

walro replied:
I think that will lead to confusion regarding what is returned from the method. Otherwise I like the proposed change.

Maybe we could refactor so that start and stop is handled by the tests. We might need to take alook at how we can implement a non-blocking behavior in Ecco then, though.

roback replied:
> I think that will lead to confusion regarding what is returned from the method. Otherwise I like the proposed change.

`start_client_get_first_row_event_then_stop_client_and_then_return_the_event` 

Much better :P

walro replied:
Hehe, I think we can leave it as-is for now.

dentarg commented:
Nice work here @roback :sparkles:

walro commented on spec/database_helper.rb:
```diff
@@ -0,0 +1,36 @@
+require "sequel"
+
+class DatabaseHelper
+  USER = ENV.fetch("DATABASE_USER")
```
Maybe we could default to root and no pass (possibly a default test-db as well), like we do in the documentdb tests. If we can live without `.env` for testing I think that would be great. Not sure what the common Twingly opinion is though. :)

dentarg replied:
Oh yes, I completely agree. `dotenv` / `.env` for tests is no good, can in fact be dangerous because sometimes (only sometimes) we populate our local machines with... let's says "non local values" to do stuff.

roback replied:
Sure, I'll try to setup a vagrant machine and use user/pass from there as the default in the tests.

roback committed:
Test that the correct values are included in the row events

Also move the TestHelper code into let block so its doesn’t have to be
repeated.

roback committed:
Vagrant machine for testing

* Installs mysql
* Enables mysql replication
* Creates database for tests

Set environment variable defaults so you don’t have to use dotenv
(https://github.com/twingly/ecco/pull/3#discussion_r44518684)

roback committed:
Add short vagrant explanation to readme

Didn’t know where to put it. The readme probably needs some
restructuring later.

roback committed:
Don't run integration tests by default

roback committed:
Run all specs on travis

Forgot this in bd583f8af26d9046ce2356aea9e120832a93cafe

roback commented:
> add bundle exec :)

Thanks :)

roback committed:
Fix tests on travis

Travis runs `bundle exec rake` by default, which isn’t clear in their
documentation :’(

Broke it in 46d2abc1fad94da347f4cb2bebc8599705bbfa4b

roback committed:
More integration tests

* Add some describes
* Client#on_row_event
  * Logfile rotation
  * Inserts in multiple tables
* Client#on_save_position
  * Position gets incremented
  * Logfile rotation

roback committed:
Test client position and filename setters

Had to use a second ecco client to be able to get a previous binlog
position for the subject to start on.

I tried reusing the subject, but it seemed like the java client saved
its last position internally and started from it, so the tests passed
even if you didn’t call set_binlog_position.

roback committed:
Cleanup TestHelper class

roback committed:
Handle timeouts in tests

If something goes wrong and no binlog events are received,
the client will block the thread forever unless there is a timeout.

The same thing was done in twingly/twingly-amqp#22

walro commented:
Nice work! I have looked at the code here at GitHub, will try it out tomorrow!

roback commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;CLRL3ORF&#x2F;1439-sprint34-bruce-problem-vid-omstarter-och-lokal-korning"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Sprint34: Bruce - Problem vid omstarter och lokal körning</a>

walro commented:
> Nice work! I have looked at the code here at GitHub, will try it out tomorrow!

It works, but I find the logging from the mysql-binlog-connector a bit annoying. Will take a look at that.

walro committed:
Increase default Java log level to Warning for integration tests

walro committed:
Randomize test order

walro committed:
Add negative tests for #start

I think we want to encapsulate the java-exception.

walro committed:
Encapsulate java's IOException

walro commented on spec/test_helper.rb:
```diff
@@ -0,0 +1,52 @@
+  def self.start_client_in_thread_and_run_block(ecco_client)
+    Timeout.timeout(CLIENT_TIMEOUT) do
+      thread = Thread.new do |t|
+        ecco_client.start
```
If we exposed the following method from `mysql-binlog-connector` I think that we could simplify our tests: https://github.com/shyiko/mysql-binlog-connector-java/blob/7e30439ea14e39d4a78d6d3673369b1dfb7e3ad0/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L555. 

I like how this looks: https://github.com/shyiko/mysql-binlog-connector-java/blob/7e30439ea14e39d4a78d6d3673369b1dfb7e3ad0/src/test/java/com/github/shyiko/mysql/binlog/BinaryLogClientIntegrationTest.java#L117-L126

roback replied:
:+1: Then the thread and sleep calls could be removed. I'll try that and see how it looks.

dentarg replied:
> Then the thread and sleep calls could be removed

I like the sound of that

roback committed:
Add non-blocking connect method to ecco client

…and use it in tests.

The tests don’t have to run the client in separate threads anymore,
which means that they are faster as you don’t have to sleep for a
second anymore (was used to make sure the client in the thread was
started).

The default timeout is from the tests in mysql-binlog-connector-java.

Don’t know how we can get rid of the duplicated code in TestHelper so
I’ll leave that as it is for now.

walro committed:
Improve on error message

roback commented:
Tried adding a test for the blocking `client#start` method:

``` ruby
  describe "#start" do
    let(:sleep_seconds) { 3 }
    let(:seconds_blocked) do
      Benchmark.realtime { subject.start }
    end

    it "should block the current thread until its stopped" do
      event_thread = Thread.new do
        sleep sleep_seconds
        DatabaseHelper.insert(table_name, mysql_row)
      end

      # Stop client on first event
      subject.on_row_event do
        subject.stop
      end

      expect(seconds_blocked.round).to eq(sleep_seconds)

      event_thread.join
    end
  end
```

but that failed:

```
  1) Ecco::Client#start should block the current thread until its stopped
     Failure/Error: expect(seconds_blocked.round).to eq(sleep_seconds)

       expected: 3
            got: 16

       (compared using ==)
     # ./spec/integration/client_spec.rb:49:in `block in (root)'
```

It looks like the `DatabaseHelper.insert` method doesn't work correctly when using threads. I timed it and it took between `1-30` seconds to run. :confused:

walro commented:
I think we can live without the `#start` test for now, we're just delegating anyway

dentarg committed:
Add a note about Ansible when using Vagrant

dentarg committed:
Give the Vagrant box a name and hostname

Easier to manage your Vagrant machines when you have a proper name
rather than "default".

roback committed:
Tests for #start_in_thread and #stop

This documents the current behaviour. We probably don’t
want to get a timeout exception when #start_in_thread is
called multiple times in the future.

dentarg committed:
Extract optimized jruby command to helper scripts

Also use it on Travis CI, tests should be a little bit quicker.

dentarg committed:
Test with JRuby 9.0.3.0 on Travis CI

dentarg commented:
Got a failure when running locally on my iMac

```
Failures:

  1) Ecco::Client#stop when called multiple times should not throw an error
     Failure/Error: Unable to find com.github.shyiko.mysql.binlog.BinaryLogClient.waitForKeepAliveThreadToBeTerminated(com/github/shyiko/mysql/binlog/BinaryLogClient.java to read failed line

     Java::JavaLang::IllegalStateException:
       BinaryLogClient was unable to shut keep alive thread down in 6000ms
     # com.github.shyiko.mysql.binlog.BinaryLogClient.waitForKeepAliveThreadToBeTerminated(com/github/shyiko/mysql/binlog/BinaryLogClient.java:835)
     # com.github.shyiko.mysql.binlog.BinaryLogClient.disconnect(com/github/shyiko/mysql/binlog/BinaryLogClient.java:820)
     # java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:606)
     # RUBY.stop(/Users/dentarg/twingly/ruby/ecco/lib/ecco/client.rb:46)
     # RUBY.block in (root)(/Users/dentarg/twingly/ruby/ecco/spec/integration/client_spec.rb:49)
     # java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:599)

Finished in 10.8 seconds (files took 1.23 seconds to load)
19 examples, 1 failure

Failed examples:

rspec ./spec/integration/client_spec.rb:46 # Ecco::Client#stop when called multiple times should not throw an error

Randomized with seed 850

/Users/dentarg/.rubies/jruby-9.0.4.0/bin/jruby -I/Users/dentarg/.gem/jruby/2.2.2/gems/rspec-core-3.4.0/lib:/Users/dentarg/.gem/jruby/2.2.2/gems/rspec-support-3.4.0/lib /Users/dentarg/.gem/jruby/2.2.2/gems/rspec-core-3.4.0/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb failed
```

dentarg commented:
Going to run `while true ; do ; sleep 1; bin/all_specs ; done` for a while

roback commented:
> Going to run while true ; do ; sleep 1; bin/all_specs ; done for a while

:+1:

dentarg commented:
5 failures during 12m 30s

Going to try jruby 9.0.3.0

dentarg commented:
During 46m 3s of `while true ; do ; sleep 60; bin/all_specs ; done` (note 60 seconds sleep)

Happened 1 time

```
Failures:

  1) Ecco::Client#stop when called multiple times should not throw an error
     Failure/Error: Unable to find com.github.shyiko.mysql.binlog.BinaryLogClient.waitForKeepAliveThreadToBeTerminated(com/github/shyiko/mysql/binlog/BinaryLogClient.java to read failed line

     Java::JavaLang::IllegalStateException:
       BinaryLogClient was unable to shut keep alive thread down in 6000ms
     # com.github.shyiko.mysql.binlog.BinaryLogClient.waitForKeepAliveThreadToBeTerminated(com/github/shyiko/mysql/binlog/BinaryLogClient.java:835)
     # com.github.shyiko.mysql.binlog.BinaryLogClient.disconnect(com/github/shyiko/mysql/binlog/BinaryLogClient.java:820)
     # java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:606)
     # RUBY.stop(/Users/dentarg/twingly/ruby/ecco/lib/ecco/client.rb:46)
     # RUBY.block in (root)(/Users/dentarg/twingly/ruby/ecco/spec/integration/client_spec.rb:49)
     # java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:599)

Finished in 10.52 seconds (files took 1.13 seconds to load)
19 examples, 1 failure

Failed examples:

rspec ./spec/integration/client_spec.rb:46 # Ecco::Client#stop when called multiple times should not throw an error

Randomized with seed 14558

/Users/dentarg/.rubies/jruby-9.0.3.0/bin/jruby -I/Users/dentarg/.gem/jruby/2.2.2/gems/rspec-core-3.4.0/lib:/Users/dentarg/.gem/jruby/2.2.2/gems/rspec-support-3.4.0/lib /Users/dentarg/.gem/jruby/2.2.2/gems/rspec-core-3.4.0/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb failed
/Users/dentarg/.rubies/jruby-9.0.3.0/bin/jruby -I/Users/dentarg/.gem/jruby/2.2.2/gems/rspec-core-3.4.0/lib:/Users/dentarg/.gem/jruby/2.2.2/gems/rspec-support-3.4.0/lib /Users/dentarg/.gem/jruby/2.2.2/gems/rspec-core-3.4.0/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb

Randomized with seed 43956
```

twingly-mob committed:
Remove unneeded semi-colon

Copy paste failure from Java example

twingly-mob committed:
Use safer java_import over import

From
https://github.com/jruby/jruby/wiki/CallingJavaFromJRuby#using-a-java-cl
ass-without-the-full-path-name

twingly-mob committed:
Import java.io.IOException class

Makes for shorter rescue

twingly-mob committed:
Remove unreliable tests

We don’t think that mysql-binlog-connector is designed to handle
multiple disconnects or multiple connects.

Moreover, we did not test any Ecco specific functionality.

twingly-mob committed:
Simplify TestHelper a little bit

twingly-mob commented:
We are now happy.
